File: api-quarkus/src/main/java/br/ufrn/musi/dominio/Dominio.java
package br.ufrn.musi.dominio;

import java.time.Instant;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.regex.Pattern;

/**
 * O domínio do MUSI em Java.
 *
 * É a mesma modelagem que existe em Kotlin, em `shared/`.
 *
 * A duplicação é proposital — ver docs/decisoes/0001-stacks-e-estrutura.md. A comparação
 * entre as duas linguagens é um importante conteúdo da disciplina, e importar o módulo Kotlin
 * diretamente anularia isso. O que impede as duas de divergirem são os casos de teste
 * compartilhados de `contratos/exemplos/`.
 *
 * Vale abrir este arquivo ao lado de `shared/src/commonMain/.../Dominio.kt`.
 */
public final class Dominio {
    private Dominio() {}

    /** Um par (dimensão, valor). A lista de dimensões não é fixa — ADR-0002. */
    public record Faceta(String dimensao, String valor) {}

    /**
     * `mbid` e `mbidComposicao` são a identidade global, preenchida na conciliação com
     * o MusicBrainz — ADR-0003. Nulos até a obra ser conciliada; a busca não depende deles.
     */
    public record Obra(String id, String titulo, String artista, int ano,
                       List<Faceta> facetas, String mbid, String mbidComposicao) {}

    /**
     * A busca é uma árvore — ADR-0002.
     *
     * `sealed` em Java 21 corresponde à `sealed interface` de Kotlin. Com pattern
     * matching em `switch`, o compilador exige que todos os casos sejam tratados —
     * a mesma garantia que o `when` exaustivo dá do outro lado.
     */
    public sealed interface Filtro permits Tem, Ou, E, Exceto, Ate {}

    public record Tem(String dimensao, String valor) implements Filtro {}
    public record Ou(List<Filtro> opcoes)            implements Filtro {}
    public record E(List<Filtro> exigencias)         implements Filtro {}
    public record Exceto(Filtro filtro)              implements Filtro {}
    public record Ate(int ano)                       implements Filtro {}

    /**
     * Avalia o filtro contra uma obra.
     *
     * Não há `default:`, e essa é uma escolha deliberada: o switch é exaustivo sobre
     * a interface selada, então acrescentar um construtor faz este método deixar de
     * compilar, e o compilador indica onde falta tratamento.
     */
    public static boolean satisfaz(Obra obra, Filtro filtro) {
        return switch (filtro) {
            case Tem t -> obra.facetas().stream()
                              .anyMatch(f -> f.dimensao().equals(t.dimensao())
                                          && f.valor().equals(t.valor()));
            case Ou o     -> o.opcoes().stream().anyMatch(f -> satisfaz(obra, f));
            case E e      -> e.exigencias().stream().allMatch(f -> satisfaz(obra, f));
            case Exceto x -> !satisfaz(obra, x.filtro());
            case Ate a    -> obra.ano() <= a.ano();
        };
    }

    public static String descrever(Filtro filtro) {
        return switch (filtro) {
            case Tem t    -> t.dimensao() + "=" + t.valor();
            case Ou o     -> juntar(o.opcoes(), " ou ");
            case E e      -> juntar(e.exigencias(), " e ");
            case Exceto x -> "não " + descrever(x.filtro());
            case Ate a    -> "ano<=" + a.ano();
        };
    }

    private static String juntar(List<Filtro> filtros, String separador) {
        var sb = new StringBuilder("(");
        for (int i = 0; i < filtros.size(); i++) {
            if (i > 0) sb.append(separador);
            sb.append(descrever(filtros.get(i)));
        }
        return sb.append(")").toString();
    }

    /**
     * A ordenação é sempre declarada por quem consulta — ADR-0002.
     */
    public enum Ordenacao { TITULO, ARTISTA, ANO_CRESCENTE, ANO_DECRESCENTE }

    /**
     * Uma faceta atribuída a uma obra por um curador identificável — ver docs/DOMINIO.md.
     * Anotações divergentes de curadores diferentes coexistem.
     */
    public record Anotacao(long id, String obraId, Faceta faceta, String curador, Instant criadoEm) {}

    /*
     * As regras de forma do domínio, as mesmas de contratos/obra.schema.json e de
     * `Regras` no Dominio.kt. Devolvem a lista de violações em vez de lançar exceção:
     * quem valida uma entrada quer todas as mensagens de uma vez.
     */
    public static final Pattern TERMO = Pattern.compile("^[a-z0-9]+(-[a-z0-9]+)*$");
    public static final int ANO_MINIMO = 1877;   // o fonógrafo
    public static final int ANO_MAXIMO = 2100;
    public static final int TAMANHO_MAXIMO_TEXTO = 300;
    public static final int TAMANHO_MAXIMO_TERMO = 60;
    public static final int TAMANHO_MAXIMO_CURADOR = 120;

    public static List<String> violacoes(Faceta f) {
        var v = new ArrayList<String>();
        if (!termoValido(f.dimensao()))
            v.add("dimensão `" + f.dimensao() + "` fora do padrão: minúsculas, sem acento, com hífen");
        if (!termoValido(f.valor()))
            v.add("valor `" + f.valor() + "` fora do padrão: minúsculas, sem acento, com hífen");
        return v;
    }

    /** Violações dos campos de uma obra, antes de ela ter identidade. */
    public static List<String> violacoesDaObra(String titulo, String artista, int ano, List<Faceta> facetas) {
        var v = new ArrayList<String>();
        if (titulo.isBlank()) v.add("título vazio");
        if (titulo.length() > TAMANHO_MAXIMO_TEXTO) v.add("título com mais de " + TAMANHO_MAXIMO_TEXTO + " caracteres");
        if (artista.isBlank()) v.add("artista vazio");
        if (artista.length() > TAMANHO_MAXIMO_TEXTO) v.add("artista com mais de " + TAMANHO_MAXIMO_TEXTO + " caracteres");
        if (ano < ANO_MINIMO || ano > ANO_MAXIMO) v.add("ano " + ano + " fora de " + ANO_MINIMO + ".." + ANO_MAXIMO);
        facetas.forEach(f -> v.addAll(violacoes(f)));
        if (new HashSet<>(facetas).size() != facetas.size()) v.add("faceta repetida");
        return v;
    }

    public static List<String> violacoesDoCurador(String curador) {
        var v = new ArrayList<String>();
        if (curador.isBlank()) v.add("curador vazio: toda anotação é assinada");
        if (curador.length() > TAMANHO_MAXIMO_CURADOR) v.add("curador com mais de " + TAMANHO_MAXIMO_CURADOR + " caracteres");
        return v;
    }

    private static boolean termoValido(String termo) {
        return TERMO.matcher(termo).matches() && termo.length() <= TAMANHO_MAXIMO_TERMO;
    }
}

File: shared/src/commonMain/kotlin/br/ufrn/musi/dominio/Dominio.kt
package br.ufrn.musi.dominio

import kotlin.time.Instant

/**
 * O domínio do MUSI em Kotlin.
 *
 * DIM0524 — Sistemas para Dispositivos Móveis.
 *
 * Este arquivo fica em `commonMain`: compila para Android, iOS e desktop, e não importa
 * Compose nem Ktor. A camada de domínio não conhece a interface nem a rede.
 * A regra está no `docs/STACK.md` da disciplina e é verificada pelo compilador: um import
 * de Compose aqui faz a compilação falhar.
 *
 * Ver docs/decisoes/0002 e 0003.
 */

/** Um par (dimensão, valor). A lista de dimensões não é fixa — ADR-0002. */
data class Faceta(val dimensao: String, val valor: String)

data class Obra(
    val id: String,
    val titulo: String,
    val artista: String,
    val ano: Int,
    val facetas: List<Faceta>,
    // Identidade global, preenchida na conciliação com o MusicBrainz — ADR-0003.
    // Nulo até a obra ser conciliada; a busca não depende deles.
    val mbid: String? = null,             // a gravação (recording)
    val mbidComposicao: String? = null,   // a composição (work), compartilhada por regravações
)

/**
 * A busca é uma árvore — ADR-0002.
 *
 * `sealed` indica que o compilador conhece todas as implementações, então um `when`
 * sobre Filtro é exaustivo: acrescentar um construtor faz a compilação falhar em todos
 * os pontos que precisam ser atualizados.
 *
 * Go não tem união etiquetada, e por isso não oferece essa verificação.
 */
sealed interface Filtro {
    data class Tem(val dimensao: String, val valor: String) : Filtro
    data class Ou(val opcoes: List<Filtro>) : Filtro
    data class E(val exigencias: List<Filtro>) : Filtro
    data class Exceto(val filtro: Filtro) : Filtro
    data class Ate(val ano: Int) : Filtro
}

/** Avalia o filtro contra uma obra. A recursão percorre a árvore. */
fun Obra.satisfaz(f: Filtro): Boolean = when (f) {
    is Filtro.Tem    -> facetas.any { it.dimensao == f.dimensao && it.valor == f.valor }
    is Filtro.Ou     -> f.opcoes.any { satisfaz(it) }
    is Filtro.E      -> f.exigencias.all { satisfaz(it) }
    is Filtro.Exceto -> !satisfaz(f.filtro)
    is Filtro.Ate    -> ano <= f.ano
}

/** Descrição legível do filtro, para depuração e para a interface. */
fun descrever(f: Filtro): String = when (f) {
    is Filtro.Tem    -> "${f.dimensao}=${f.valor}"
    is Filtro.Ou     -> f.opcoes.joinToString(" ou ", "(", ")") { descrever(it) }
    is Filtro.E      -> f.exigencias.joinToString(" e ", "(", ")") { descrever(it) }
    is Filtro.Exceto -> "não ${descrever(f.filtro)}"
    is Filtro.Ate    -> "ano<=${f.ano}"
}

/**
 * A ordenação é sempre declarada por quem consulta — ADR-0002.
 * Não há um valor `POR_RELEVANCIA`, e essa ausência é deliberada.
 */
enum class Ordenacao { TITULO, ARTISTA, ANO_CRESCENTE, ANO_DECRESCENTE }

fun List<Obra>.buscar(filtro: Filtro, ordenacao: Ordenacao = Ordenacao.TITULO): List<Obra> =
    filter { it.satisfaz(filtro) }.sortedWith(
        when (ordenacao) {
            Ordenacao.TITULO          -> compareBy { it.titulo }
            Ordenacao.ARTISTA         -> compareBy { it.artista }
            Ordenacao.ANO_CRESCENTE   -> compareBy { it.ano }
            Ordenacao.ANO_DECRESCENTE -> compareByDescending { it.ano }
        }
    )

/**
 * Uma faceta atribuída a uma obra por um curador identificável — ver docs/DOMINIO.md.
 *
 * Anotações divergentes de curadores diferentes coexistem: o catálogo registra quem afirma
 * o quê, sem arbitrar entre as afirmações. `criadoEm` é o instante da anotação; a
 * representação (ISO-8601) é decisão de quem serializa.
 */
data class Anotacao(
    val id: Long,
    val obraId: String,
    val faceta: Faceta,
    val curador: String,
    val criadoEm: Instant,
)

/**
 * As regras de forma do domínio, as mesmas de contratos/obra.schema.json.
 *
 * Devolvem a lista de violações em vez de lançar exceção: quem valida uma entrada quer
 * todas as mensagens de uma vez, não a primeira.
 */
object Regras {
    /** Minúsculas, sem acento, com hífen — docs/GLOSSARIO.md. */
    val TERMO = Regex("^[a-z0-9]+(-[a-z0-9]+)*$")
    const val ANO_MINIMO = 1877       // o fonógrafo
    const val ANO_MAXIMO = 2100
    const val TAMANHO_MAXIMO_TEXTO = 300
    const val TAMANHO_MAXIMO_TERMO = 60
    const val TAMANHO_MAXIMO_CURADOR = 120
}

fun Faceta.violacoes(): List<String> = buildList {
    if (!Regras.TERMO.matches(dimensao) || dimensao.length > Regras.TAMANHO_MAXIMO_TERMO)
        add("dimensão `$dimensao` fora do padrão: minúsculas, sem acento, com hífen")
    if (!Regras.TERMO.matches(valor) || valor.length > Regras.TAMANHO_MAXIMO_TERMO)
        add("valor `$valor` fora do padrão: minúsculas, sem acento, com hífen")
}

/** Violações dos campos de uma obra, antes de ela ter identidade. */
fun violacoesDaObra(titulo: String, artista: String, ano: Int, facetas: List<Faceta>): List<String> = buildList {
    if (titulo.isBlank()) add("título vazio")
    if (titulo.length > Regras.TAMANHO_MAXIMO_TEXTO) add("título com mais de ${Regras.TAMANHO_MAXIMO_TEXTO} caracteres")
    if (artista.isBlank()) add("artista vazio")
    if (artista.length > Regras.TAMANHO_MAXIMO_TEXTO) add("artista com mais de ${Regras.TAMANHO_MAXIMO_TEXTO} caracteres")
    if (ano !in Regras.ANO_MINIMO..Regras.ANO_MAXIMO) add("ano $ano fora de ${Regras.ANO_MINIMO}..${Regras.ANO_MAXIMO}")
    facetas.forEach { addAll(it.violacoes()) }
    if (facetas.size != facetas.toSet().size) add("faceta repetida")
}

fun violacoesDoCurador(curador: String): List<String> = buildList {
    if (curador.isBlank()) add("curador vazio: toda anotação é assinada")
    if (curador.length > Regras.TAMANHO_MAXIMO_CURADOR) add("curador com mais de ${Regras.TAMANHO_MAXIMO_CURADOR} caracteres")
}
